src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -408,28 +405,34 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
         offset.x = clampedOverhang.left - clampedOverhang.right;
         offset.y = clampedOverhang.top - clampedOverhang.bottom;
 
-        if (actualSize.x > targetSize.x)
+        if (actualSize.x > layoutBox.x)
         {
             scalingRequired = true;
             offset.x = (overhang.left - overhang.right) * 0.5f;
-            scale.x = targetSize.x / actualSize.x;
+            scale.x = (layoutBox.x - _r.dipPerPixel) / actualSize.x;
             scale.y = scale.x;
         }
-        if (actualSize.y > targetSize.y)
+        if (actualSize.y > layoutBox.y)
         {
             scalingRequired = true;
             offset.y = (overhang.top - overhang.bottom) * 0.5f;
-            scale.x = std::min(scale.x, targetSize.y / actualSize.y);
+            scale.x = std::min(scale.x, (layoutBox.y - _r.dipPerPixel) / actualSize.y);
             scale.y = scale.x;
         }
 
         offset.x *= scale.x;
         offset.y *= scale.y;
 
-        // Due to D2D1_DRAW_TEXT_OPTIONS_NO_SNAP our baseline won't get
-        // pixel aligned anymore and so we have to do it ourselves here.
-        offset.x = roundf(offset.x * _r.pixelPerDIP) * _r.dipPerPixel;
-        offset.y = roundf(offset.y * _r.pixelPerDIP) * _r.dipPerPixel;
+        // As explained below, we use D2D1_DRAW_TEXT_OPTIONS_NO_SNAP to prevent a weird issue with baseline snapping.
+        // But we do want it technically, so this re-implements baseline snapping... I think?
+        // It calculates the new `baseline` height after transformation by `scale.y` relative to the center point `halfSize.y`.
+        //
+        // This works even if `scale.y == 1`, because then `baseline == baselineInDIP + offset.y` and `baselineInDIP`

```

Comment:
to some extent i just have to, like... trust your math lol

Reply:
Yeah, this is really hard to read:
```
baseline = halfSize.y + (_r.fontMetrics.baselineInDIP + offset.y - halfSize.y) * scale.y
```

If it helps here's an image explanation:
![image](https://user-images.githubusercontent.com/2256941/181860430-5beed2f0-df00-44cb-8f44-13c7111bec10.png)

`offset.y` shifts the glyph (and thus the baseline) up and down as needed to fit the glyph into the cell. So it's a simple `+ offset.y` and you arrive at my complex line of code there.